drivers/cyw43/cywbt.c

```diff
@@ -29,42 +29,48 @@
 
 #include "py/runtime.h"
 #include "py/mphal.h"
-#include "pin_static_af.h"
-#include "uart.h"
 #include "extmod/mpbthci.h"
 
 #if MICROPY_PY_NETWORK_CYW43
 
 #include "lib/cyw43-driver/src/cyw43_config.h"
 #include "lib/cyw43-driver/firmware/cyw43_btfw_4343A1.h"
 
-// Provided by the port.
-extern pyb_uart_obj_t mp_bluetooth_hci_uart_obj;
-
 // Provided by the port, and also possibly shared with the stack.
 extern uint8_t mp_bluetooth_hci_cmd_buf[4 + 256];
 
 /******************************************************************************/
 // CYW BT HCI low-level driver
 
+#ifdef CYW43_PIN_BT_CTS
+// This code is not portable and currently only builds on stm32 port.
+
+#include "pin_static_af.h"
+#include "uart.h"
+
+// Provided by the port.
+extern pyb_uart_obj_t mp_bluetooth_hci_uart_obj;
+
 STATIC void cywbt_wait_cts_low(void) {
-    mp_hal_pin_config(pyb_pin_BT_CTS, MP_HAL_PIN_MODE_INPUT, MP_HAL_PIN_PULL_UP, 0);
+    mp_hal_pin_config(CYW43_PIN_BT_CTS, MP_HAL_PIN_MODE_INPUT, MP_HAL_PIN_PULL_UP, MP_HAL_PIN_AF_GPIO);
     for (int i = 0; i < 200; ++i) {
-        if (mp_hal_pin_read(pyb_pin_BT_CTS) == 0) {
+        if (mp_hal_pin_read(CYW43_PIN_BT_CTS) == 0) {
             break;
         }
         mp_hal_delay_ms(1);
     }
-    mp_hal_pin_config_alt(pyb_pin_BT_CTS, MP_HAL_PIN_MODE_ALT, MP_HAL_PIN_PULL_UP, AF_FN_UART, mp_bluetooth_hci_uart_obj.uart_id);
+    mp_hal_pin_config_alt(CYW43_PIN_BT_CTS, MP_HAL_PIN_MODE_ALT,
+        MP_HAL_PIN_PULL_UP, AF_FN_UART, mp_bluetooth_hci_uart_obj.uart_id);
 }
+#endif
 
 STATIC int cywbt_hci_cmd_raw(size_t len, uint8_t *buf) {
-    uart_tx_strn(&mp_bluetooth_hci_uart_obj, (void*)buf, len);
+    mp_bluetooth_hci_uart_write((void *)buf, len);
     for (int i = 0; i < 6; ++i) {
-        while (!uart_rx_any(&mp_bluetooth_hci_uart_obj)) {
+        while (!mp_bluetooth_hci_uart_any()) {
             MICROPY_EVENT_POLL_HOOK
         }
-        buf[i] = uart_rx_char(&mp_bluetooth_hci_uart_obj);
+        buf[i] = mp_bluetooth_hci_uart_readchar();
```

Comment:
Would be simpler to write this as:
```c
int c;
while ((c = mp_bluetooth_hci_uart_readchar()) == -1) {
    MICROPY_EVENT_POLL_HOOK
}
buf[i] = c;
```

Reply:
But `mp_bluetooth_hci_uart_readchar()` calls `uart_rx_any()` before it tries to read (on stm32 at least).  So in the PR as it currently stands, any() is called at least twice for each character read.

The version I propose above should be equivalent to the original code in terms of the ordering and number of calls to `uart_rx_any()` and `uart_rx_char()`.  Unless I missed something?

Reply:
> But `mp_bluetooth_hci_uart_readchar()` calls `uart_rx_any()` before it tries to read

Oh okay I see your point, changed.